### PR TITLE
Planner: support window

### LINKED ISSUE #4739: interpreter planner refactor
#### background
Currently tiflash's interpreter is designed based on `DAGQueryBlock`.
`DAGQueryBlock` is designed to coprocess read, and it can handle simple queries well.
However, there is no way to handle the complex query of MPP, so the `DAGQueryBlock` tree is introduced, which is a compromise that makes the code more complicated and affects the subsequent iterative development.
We need to refactor the interpreter code in a plan visit way.

#### design
Therefore we introduce a new abstraction layer called `PhysicalPlan` to replace `DAGQueryBlock`.
It will be responsible for the processing of `ExpressionAction` and the generation of `BlockInputStream`.
We call this organization of code the planner.

```
DAGRequest (executor tree or executor array).
     |
     |
     v
physical plan <---+
     |            |
     |            | optimize
     |            |
     +------------+   
     |
     |
     v
BlockInputStream
```

#### why planner
Most databases are designed based on planner. 
- Friendly for database developers to understand code.
- We can improve tiflash with a general approach from the DB area, such as pipeline mode.

#### todo list
- [x] public desgin doc
    - #4743
- [x] Small refactoring of Interpreter before planner refactoring 
     - #4118 
- [x] interpreter test framework 
    - #4609 
- [x] planner refactory
    - [x] poc #4601 
    - [x] Introduce `PhysicalPlan`/`planner`, add switch and set it to planner by default
        - https://github.com/pingcap/tiflash/pull/4820 
    - [x] Replace all operators with `PhysicalPlan`
        - [x] common plans: projection/agg/filter/exchange_(sender/receiver)/limit/topn
            - https://github.com/pingcap/tiflash/pull/5143
        - [x] window function
            - https://github.com/pingcap/tiflash/pull/5363
        - [x] join
            - https://github.com/pingcap/tiflash/pull/5320
        - [x] table scan
            - https://github.com/pingcap/tiflash/pull/5335
    - [x] remove `DAGQueryBlock`
            - https://github.com/pingcap/tiflash/pull/5381
- [x] merge to master
    - feature branch: https://github.com/pingcap/tiflash/tree/planner_refactory
    - https://github.com/pingcap/tiflash/pull/5493
    - https://github.com/pingcap/tiflash/pull/5321

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4739

Problem Summary:

### What is changed and how it works?
- move `enableFineGrainedShuffleExtraInfo` and `enableFineGrainedShuffle` from `DAGContext` to `FineGrainedShuffle.h`.
- Let `PhysicalExchangeSender` and `PhysicalExchangeReceiver` support fine grained shuffle.
- Add `PhysicalWindow` and `PhysicalWindowSort`.
- Add `PhysicalPlan.disableRestoreConcurrency` to avoid `restoreConcurrency` after window_sort that disable FineGrainedShuffle.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Planner/plans/PhysicalWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+PhysicalPlanNodePtr PhysicalWindow::build(
+    const Context & context,
+    const String & executor_id,
+    const LoggerPtr & log,
+    const tipb::Window & window,
+    const FineGrainedShuffle & fine_grained_shuffle,
+    const PhysicalPlanNodePtr & child)
+{
+    assert(child);
+    child->disableRestoreConcurrency();
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/plans/PhysicalWindow.cpp:37]: child is `PhysicalSource <-- PhysicalFinalProjection <-- PhysicalWindowSort` and not `PhysicalWindowSort`.
So we need call `disableRestoreConcurrency` in `PhysicalWindowSort` and `PhysicalFinalProjection `.
After DAGQueryBlock removed, the plan tree will like `PhysicalWindow <-- PhysicalSort`

File: dbms/src/Flash/Planner/plans/PhysicalExchangeSender.cpp [hunk 2]
```diff
@@ -57,21 +59,45 @@ void PhysicalExchangeSender::transformImpl(DAGPipeline & pipeline, Context & con
 
     /// todo support fine grained shuffle
```
Comment (guo-shaoge) [inline dbms/src/Flash/Planner/plans/PhysicalExchangeSender.cpp:60]:
```suggestion
```

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Flash/Planner/plans/PhysicalWindow.cpp (hunk 1 above)
Comment (mengxin9014) [reply dbms/src/Flash/Planner/plans/PhysicalWindow.cpp:37]: Why If child is PhysicalSource <-- PhysicalFinalProjection <-- PhysicalWindowSort and not PhysicalWindowSort, we need call disableRestoreConcurrency in PhysicalWindowSort and PhysicalFinalProjection.

**Comment by mengxin9014** [comment]
I suggest adding a few unit tests with multi-stream.

**Comment by mengxin9014** [comment]
others LGTM.

**Comment by SeaRise** [comment]
> I suggest adding a few unit tests with multi-stream.

wait for https://github.com/pingcap/tiflash/pull/5464 merged and merge to planner_refactory,

**Comment by SeaRise** [comment]
> > I suggest adding a few unit tests with multi-stream.
> 
> wait for #5464 merged and merge to planner_refactory,

I has port the test cases from #5464.

**Review by mengxin9014 — APPROVED** [review]

**Pull request merged**